### ISSUE TITLE
TimestampBehavior not working

### ISSUE THREAD

**Issue by i-panov** [yiisoft/yii2#15946]
The project and this model is already six months old and all worked fine before, but suddenly the TimestampBehavior broke down. The database writes all the time zeros (in those fields where there should be time). I did not change anything related to this. It feels like a crash somewhere inside the framework...

Important! I noticed in the debugger that after saving the model the fields it is not empty !!! But the database still writes zeros. It's very strange...

**Comment by i-panov** [comment]
I inform you about the bug. It's not a question!

**Comment by i-panov** [comment]
For version 2.0.15 the problem is also relevant.

**Comment by larryli** [comment]
@i-panov Please read https://github.com/yiisoft/yii2/blob/master/docs/guide/concept-behaviors.md#using-timestampbehavior-

**Comment by i-panov** [comment]
@larryli What have I done wrong? What did not specify all the fields (attributes, value, etc.)? But this is not necessarily as far as I know. As fields for dates at me are called by default created_at and updated_at.

**Comment by i-panov** [comment]
From the comment to the behaviours method:

> The return value of this method should be an array of behavior objects or configurations indexed by behavior names. A behavior configuration can be either a **string specifying the behavior class** or an array of the following structure:

**Comment by i-panov** [comment]
@FabrizioCaldarelli If you mean that I specified `class` instead of `className()`, then you yourself deprecated this method!

> @deprecated since 2.0.14. On PHP >=5.5, use `::class` instead.

**Comment by FabrizioCaldarelli** [comment]
No, I refer to structure. You used 'timestamp' key and TimestampBehavior::class value, I posted an array with only TimestampBehavior::className() value or 'class' key and TimestampBehavior::className() value.

**Comment by i-panov** [comment]
@FabrizioCaldarelli Once again I repeat:

> From the comment to the behaviours method:
The return value of this method should be an array of behavior objects or configurations indexed by behavior names. A behavior configuration can be either a **string specifying the behavior class** or an array of the following structure:

So, judging by your documentation, I can just use the class name instead of an array. And the `timestamp` key is needed for the detachBehavior capability.

**Comment by larryli** [comment]
@i-panov Do you use `beforeSave` function in `Article` like this:

```php
    public function beforeSave($insert)
    {
        // @bug: do not call parent::beforeSave($insert)
        return true;
    }
```

**Comment by i-panov** [comment]
@larryli No, I do not use beforeSave.

**Comment by larryli** [comment]
@i-panov If not, you can make a log in `Article::beforeSave `:

```php
 public function beforeSave($insert)
    {
        $return =parent::beforeSave($insert);
        Yii::info("created_at: {$model->created_at}, updated_at: {$model->updated_at}");
        return $return;
    }
```

**Comment by i-panov** [comment]
@larryli As I said before

> Important! I noticed in the debugger that after saving the model the fields it is not empty !!! But the database still writes zeros. It's very strange...

`created_at: 1521796945, updated_at: 1521796945`

**Comment by berosoboy** [comment]
What is your DB column data type?

**Comment by SilverFire** [comment]
@i-panov We have a [unit test](https://github.com/yiisoft/yii2/blob/master/tests/framework/behaviors/TimestampBehaviorTest.php) that verifies, that behavior works as it should. Could you change it so it fails?

**Comment by i-panov** [comment]
@larryli And in `beforeSave` and in` afterSave`, the fields `created_at` and` updated_at` are not empty. I'm sure that these fields are not rewritten anywhere (search PHPStorm does not find such places).

@berosoboy `int`

**Comment by berosoboy** [comment]
Enable db profiling and see the generated INSERT/UPDATE query in debug panel.

**Comment by i-panov** [comment]
@berosoboy I have a yii-debug-toolbar and there are all the logs. Now I looked at the insert query: there are no fields created_at and updated_at. At all. I do not know how to explain it. In beforeSave they are setted, but in the query they are not...

**Comment by berosoboy** [comment]
Do this: 
In your beforeSave, check the result of `$this->getDirtyAttributes();`

**Comment by i-panov** [comment]
FUUUUUUUUUUCK!!!!! I think I understood what's wrong ... The problem was again in your unobvious infrastructure! I called the `save (true, ['name', 'category_id', 'user_id', 'status_id'])` method. But I somehow thought that this array specifies which attributes should be validated and which ones do not! And now I read that it indicates which attributes should be saved ...

**Comment by i-panov** [comment]
Thank you all for your help. :)

**Comment by berosoboy** [comment]
You're welcome :D

**Comment by i-panov** [comment]
Simply there the first parameter goes the flag runValidation and by itself the conclusion suggests that the second parameter specifies attributes for validation ... This is much more logical. I would do so in your place. And what for to specify attributes which it is necessary to save I generally do not know ... After all, they can simply not be setted.

**Comment by berosoboy** [comment]
The `save()` method is intended to save the model. In my opinion the method params are OK.
